### DISCUSSION TITLE
Enforcing non-negative gene expression dynamics

### DISCUSSION THREAD

**Discussion by anshu957** [aristoteleo/dynamo-release#445] [General]
**_Objective_**: Predict gene expression dynamics over time in the `gene space` by integrating the vector field. 

**_Current approach_**: Learn vector field in `PCA space`,  integrate the trajectory forward in time in `PCA space` and then project the gene expression (for all time points) back to gene expression space. 

**_Issue_**: Is there a way to enforce/constraint the learned vector field function so that the solution is non-negative? I know `PCA space` is an abstract space so probably it is not possible there. But integrating forward with some genes going negative may result in inaccurate predictions.  

**_Solution explored_**: Learn the vector field in `gene space`. However, learning in the `gene space` (let's say with ~2k genes) does not result in a good vector field re-construction despite trying multiple values of `M` (number of basis function). Even if the re-construction is good, how do we enforce positveness of the solution?

Any recommendation or suggestion?

**Comment by Xiaojieqiu** [comment]
@anshu957 this is an excellent question and actually can be one part of a research problem. To solve it, we may need to design some new ways for vector field reconstruction. A naive approach to solve this can be something like the following (I am using the ode_bifur2gene as an example): 

```
from scipy.integrate import odeint

velocity_func = lambda y, t: (y>0) * dyn.sim.ode_bifur2genes(y, a=[1, 1], b=[1, 1], S=[0.5, 0.5], K=[0.5, 0.5], m=[4, 4], n=[4, 4], gamma=[1, 1])

X = []
for t in t_list:
    true_x0 = numpy.random.uniform(low=0.0, high=4.0, size=(num_of_samples,2))
    X_tmp = np.array([ odeint(velocity_func, true_x0[i], np.array([0,t]) ) for i in range(num_of_samples) ])[:,-1]
    X.append(X_tmp)
```

The key above is to use `(y > 0) *` which will make sure the integrated values always positive
